### PR TITLE
Enable compression for multinode

### PR DESCRIPTION
**Pull request by JLockerman**
This requires a custom compression job and some other workarounds for current multinode limitations.

### CODE REVIEW

File: pkg/pgmodel/migrations/sql/idempotent/base.sql [hunk 1]
```diff
@@ -144,6 +141,26 @@ COMMENT ON PROCEDURE SCHEMA_CATALOG.finalize_metric_creation()
...
+    tables TEXT[];
+BEGIN
+    IF SCHEMA_CATALOG.is_timescaledb_installed() AND SCHEMA_CATALOG.is_multinode() THEN
+        SELECT array_agg(format('%I', table_name))
+        FROM SCHEMA_CATALOG.metric
+        WHERE NOT creation_completed
+        INTO tables;
+    END IF;
+
+    RETURN tables;
```
Comment (cevian) [inline pkg/pgmodel/migrations/sql/idempotent/base.sql:160]: is this guaranteed empty at initialization?

File: pkg/pgmodel/migrations/sql/idempotent/base.sql [hunk 2]
```diff
@@ -1078,9 +1097,27 @@ BEGIN
 
         --chunks where the end time is before now()-1 hour will be compressed
         IF SCHEMA_CATALOG.get_timescale_major_version() >= 2 THEN
-            PERFORM add_compression_policy(format('SCHEMA_DATA.%I', metric_table_name), INTERVAL '1 hour');
+            IF SCHEMA_CATALOG.is_multinode() THEN
```
Comment (cevian) [inline pkg/pgmodel/migrations/sql/idempotent/base.sql:1100]: should we only do this for multinode? My feeling is do this for all of 2.0

File: pkg/pgmodel/migrations/sql/idempotent/base.sql [hunk 3]
```diff
@@ -1687,12 +1724,20 @@ DECLARE
     bgw_job_id int;
 BEGIN
     IF SCHEMA_CATALOG.get_timescale_major_version() >= 2 THEN
-        SELECT p.id INTO bgw_job_id
-        FROM _timescaledb_config.bgw_job p
-        INNER JOIN _timescaledb_catalog.hypertable h ON (h.id = p.hypertable_id)
-        WHERE p.proc_name = 'policy_compression'
-          AND h.schema_name = 'SCHEMA_DATA'
-          AND h.table_name = ht_table;
+        IF SCHEMA_CATALOG.is_multinode() THEN
```
Comment (cevian) [inline pkg/pgmodel/migrations/sql/idempotent/base.sql:1727]: again I don't think just multinode

File: pkg/pgmodel/migrations/sql/idempotent/base.sql [hunk 4]
```diff
@@ -1712,48 +1757,102 @@ END
...
+AS $proc$
+BEGIN
+    -- in early versions of timescale multinode the access node catalog does not
+    -- store whether chunks were compressed, so we need to run the actual search
+    -- for nodes in need of decompression on the data nodes
+    IF SCHEMA_CATALOG.is_multinode() THEN
+        CALL distributed_exec(
+            format(
+                $dist$ CALL do_decompress_chunks_after(%s, %s) $dist$,
+                quote_literal(metric_table), quote_literal(min_time)),
```
Comment (cevian) [inline pkg/pgmodel/migrations/sql/idempotent/base.sql:1818]:
```suggestion
                $dist$ CALL do_decompress_chunks_after(%L, %L) $dist$,
                metric_table, min_time),
```

File: pkg/pgmodel/migrations/sql/idempotent/base.sql [hunk 5]
```diff
@@ -1712,48 +1757,102 @@ END
...
+            transactional => false);
+    ELSE
+        CALL do_decompress_chunks_after(metric_table, min_time);
+    END IF;
+END
+$proc$ LANGUAGE PLPGSQL;
+
+CREATE OR REPLACE PROCEDURE SCHEMA_CATALOG.compress_old_chunks(metric_table TEXT, last_time TIMESTAMPTZ)
```
Comment (cevian) [inline pkg/pgmodel/migrations/sql/idempotent/base.sql:1826]:
```suggestion
CREATE OR REPLACE PROCEDURE SCHEMA_CATALOG.compress_old_chunks(metric_table TEXT, compress_before TIMESTAMPTZ)
```

**Review by cevian — CHANGES_REQUESTED** [review]
Some issues. Also I think we are missing tests if things are passing with these issues

File: pkg/pgmodel/migrations/sql/idempotent/base.sql (hunk 4 above)
Comment (JLockerman) [reply pkg/pgmodel/migrations/sql/idempotent/base.sql:1818]: I had a specific reason I didn't want to do that, can I remember what it was...

File: pkg/pgmodel/migrations/sql/idempotent/base.sql (hunk 2 above)
Comment (JLockerman) [reply pkg/pgmodel/migrations/sql/idempotent/base.sql:1100]: let's merge this then throw the switch

File: pkg/pgmodel/migrations/sql/idempotent/base.sql (hunk 1 above)
Comment (JLockerman) [reply pkg/pgmodel/migrations/sql/idempotent/base.sql:160]: I believe it should be NULL at init

File: pkg/pgmodel/migrations/sql/idempotent/base.sql (hunk 2 above)
Comment (cevian) [reply pkg/pgmodel/migrations/sql/idempotent/base.sql:1100]: why are we waitting?

File: pkg/pgmodel/migrations/sql/idempotent/base.sql [hunk 6]
```diff
@@ -1078,9 +1097,27 @@ BEGIN
...
+                    'SCHEMA_CATALOG.compression_job'::regproc,
+                    INTERVAL '1 hour',
+                    config => jsonb_build_object('metric_table', metric_table_name)
+                );
+            ELSE
+                PERFORM add_compression_policy(format('SCHEMA_DATA.%I', metric_table_name), INTERVAL '1 hour');
+            END IF;
         ELSE
-            PERFORM add_compress_chunks_policy(format('SCHEMA_DATA.%I', metric_table_name), INTERVAL '1 hour');
+            IF SCHEMA_CATALOG.is_multinode() THEN
```
[the remark this answers is not in the document]
Comment (cevian) [reply pkg/pgmodel/migrations/sql/idempotent/base.sql:1114]: Was a test added here?

File: pkg/pgmodel/migrations/sql/idempotent/base.sql (hunk 4 above)
Comment (cevian) [reply pkg/pgmodel/migrations/sql/idempotent/base.sql:1818]: well let's change it then unless there is a reason

File: pkg/pgmodel/migrations/sql/idempotent/base.sql [hunk 7]
```diff
@@ -1712,48 +1755,104 @@ END
...
-        COMMIT;
-    END LOOP;
-END;
-$$ LANGUAGE PLPGSQL;
+CREATE OR REPLACE PROCEDURE SCHEMA_CATALOG.compression_job(job_id int, config jsonb)
+AS $$
+    -- as of timescaledb-2.0-rc3 the is_compressed column of the chunks view is
+    -- not updated on the access node, therefore we need to one the compressor
+    -- on all the datanodes to search for uncompressed chunks
+    CALL distributed_exec(format($dist$
```
Comment (cevian) [inline pkg/pgmodel/migrations/sql/idempotent/base.sql:1855]: does this work in a non-multinode setting. If it does, then why do we need the if on line 1812? I think the code should be congruent

File: pkg/pgmodel/migrations/sql/idempotent/base.sql [hunk 8]
```diff
@@ -1712,48 +1755,104 @@ END
...
+        END LOOP;
+    END;
+    $$ LANGUAGE PLPGSQL;
+$ee$);
 
-        COMMIT;
-    END LOOP;
-END;
-$$ LANGUAGE PLPGSQL;
+CREATE OR REPLACE PROCEDURE SCHEMA_CATALOG.compression_job(job_id int, config jsonb)
```
Comment (cevian) [inline pkg/pgmodel/migrations/sql/idempotent/base.sql:1850]: let's be a bit more explicit in the naming so we don't clash with something later.
```suggestion
CREATE OR REPLACE PROCEDURE SCHEMA_CATALOG.promscale_compression_job(job_id int, config jsonb)
```

File: pkg/pgmodel/end_to_end_tests/create_test.go [hunk 9]
```diff
@@ -621,14 +615,18 @@ func TestInsertCompressed(t *testing.T) {
 		}
 		var nextStartAfter time.Time
 		var statsQuery string
-		if *useTimescale2 {
-			statsQuery = "SELECT next_start FROM timescaledb_information.job_stats WHERE hypertable_schema=$1::text AND hypertable_name = $2::text"
-			err = db.QueryRow(context.Background(), statsQuery, "prom_data", "test").Scan(&nextStartAfter)
+		if *useTimescale2 || *useMultinode {
```
Comment (cevian) [inline pkg/pgmodel/end_to_end_tests/create_test.go:618]:
```suggestion
		if *useTimescale2 {
```

File: pkg/pgmodel/end_to_end_tests/create_test.go [hunk 10]
```diff
@@ -621,14 +615,18 @@ func TestInsertCompressed(t *testing.T) {
...
+				"WHERE proc_schema = '_prom_catalog' " +
+				"AND proc_name = 'compression_job' " +
+				"AND config->>'metric_table' = $1::text"
+			err = db.QueryRow(context.Background(), statsQuery, "test").Scan(&nextStartAfter)
 		} else {
 			statsQuery = "SELECT next_start FROM timescaledb_information.policy_stats WHERE hypertable = $1::text::regclass"
 			err = db.QueryRow(context.Background(), statsQuery, pgx.Identifier{"prom_data", "test"}.Sanitize()).Scan(&nextStartAfter)
 		}
 		if err != nil {
+			time.Sleep(1 * time.Hour)
```
Comment (cevian) [inline pkg/pgmodel/end_to_end_tests/create_test.go:629]:
```suggestion
```

**Review by cevian — CHANGES_REQUESTED** [review]
Some more nits. Mostly looks good.

File: pkg/pgmodel/migrations/sql/idempotent/base.sql (hunk 7 above)
Comment (JLockerman) [reply pkg/pgmodel/migrations/sql/idempotent/base.sql:1855]: @cevian it's a nop without multinode.
I should switch it it `exec_everywhere`

File: pkg/pgmodel/migrations/sql/idempotent/base.sql (hunk 8 above)
Comment (JLockerman) [reply pkg/pgmodel/migrations/sql/idempotent/base.sql:1850]: remember, it's in `_prom_catalog` so the names are
`_prom_catalog.compression_job()` versus
`_prom_catalog. promscale_compression_job()`

File: pkg/pgmodel/migrations/sql/idempotent/base.sql (hunk 7 above)
Comment (JLockerman) [reply pkg/pgmodel/migrations/sql/idempotent/base.sql:1855]: nevermind, it's an _error_ without multinode, and 1812 is right; we shouldn't run things on the access node if it's going to be a lot of work for no benefit

Comment (JLockerman) [reply pkg/pgmodel/migrations/sql/idempotent/base.sql:1855]: (it wasn't caught because the new test only ran for multinode, not all of 2. fixed now)

**Review by cevian — APPROVED** [review]
One more nit from last time around? Otherwise LGTM

File: pkg/pgmodel/end_to_end_tests/create_test.go (hunk 9 above)
Comment (JLockerman) [reply pkg/pgmodel/end_to_end_tests/create_test.go:618]: huh, I thought I fixed that...

**Pull request merged**